### ISSUE TITLE
Unmount doesn't removed listeners from queue

### ISSUE THREAD

**Issue by gismya** [nanostores/nanostores#279]
I had an issue with removed listeners still being run, because they where queued before they where unsubscribed. Upon looking for solutions for that I see that it seems to be by design, based on one of the tests.

``` typescript
test('does not mutate listeners while change event', () => {
  let events: string[] = []
  let $store = atom<number | undefined>()

  onMount($store, () => {
    $store.set(0)
  })

  $store.listen(value => {
    events.push(`a${value}`)
    unbindB()
    $store.listen(v => {
      events.push(`c${v}`)
    })
  })

  let unbindB = $store.listen(value => {
    events.push(`b${value}`)
  })

  $store.set(1)
  deepStrictEqual(events, ['a1', 'b1'])

  $store.set(2)
  deepStrictEqual(events, ['a1', 'b1', 'a2', 'c2'])
})
```

What's the reason that listeners should not be removed if they are currently in the queue?

The use case we're working with is a class that mounts a bunch of subscribers inside. Changing one item destroys and rebuilds the class, which destroys and rebuilds the store subscribers. But if a subscriber inside the class is triggered by the same change that destroys the class, the subscriber still triggers the method inside the now destroyed class.

**Comment by gismya** [comment]
Found this out when working for a solution for it, the test I wrote was:

``` typescript
test('Unsubscribing from computed removes computed listeners from queue', () => {
  let $atom = atom(0)
  let $computed = computed($atom, value => value * 2)
  let values: (number | string)[] = []

  let unsubscribe = $computed.listen(() => {
    values.push('afterAtom')
  })
  $atom.listen(value => {
    values.push(value)
    values.push($computed.get())
    if (value > 1) {
      unsubscribe()
    }
  })
  $atom.set(1)
  deepStrictEqual(values, [1, 2, 'afterAtom'])
  $atom.set(2)
  deepStrictEqual(values, [1, 2, 'afterAtom', 2, 4])
})
```

And the solution I was testing was rewriting the listener to

``` javascript
    listen(listener, listenerLevel) {
      $atom.lc = listeners.push(listener, listenerLevel || $atom.l) / 2

      return () => {
        let index = listeners.indexOf(listener)
        let qIndex = listenerQueue.indexOf(listener)
        if (~index) {
          listeners.splice(index, 2)
          if (!--$atom.lc) $atom.off()
        }
        if (~qIndex) {
          listenerQueue.splice(index, 5)
        }
      }
    },
```

With checking the listenerqueue as well during unsubscribes.

**Comment by ai** [comment]
> What's the reason that listeners should not be removed if they are currently in the queue?

This test tests a different thing to be honest. We wrote it just to be sure that unsubscribing in the listener will not break calling other listeners (which can be done with simple `for (let i = 0; …` for-loop implementation).

So, please send PR with your changes. I agree that f we call `unsibscribe` for the listener, it listener should be never called.

**Comment by gismya** [comment]
Addressed in #280
